### DISCUSSION TITLE
Middleware conditional statements cost invocations?

### DISCUSSION THREAD

**Discussion by capJavert** [vercel/next.js#46781] [Help]
### Summary

I understand that with [matcher config](https://vercel.com/docs/concepts/functions/edge-middleware/middleware-api#match-paths-based-on-custom-matcher-config) you can limit on which paths your middleware runs and with that save on usage invocations. 

My questions is do [conditional statements](https://vercel.com/docs/concepts/functions/edge-middleware/middleware-api#match-paths-based-on-conditional-statements) also save invocations?

My gut feeling is "no" because to evaluate conditional statements you do need to run middleware function. But problem is docs show those two ways as being interchangeable even though it looks like one obviously only has more flexibilty but does not really save you on usage costs.

**Comment by guitheengineer** [comment]
Your gut feeling is right, just do a console.log inside the middleware (and obviously outside the conditional statement) and you're going to see each time is being invoked.

**Reply by capJavert** [reply]
Of course, but I thought that maybe next team did some kind of optimization there. 

If conditional statements are billed/counted always they are not much of the solution to optimization of invocations. May need to reflect this in the docs.

It would be great to be able to eg. trigger middleware based on headers/cookies etc. much like rewrites and redirects.

**Reply by guitheengineer** [reply]
Agreed, that's a great idea @capJavert. Worth proposing.
